dashboard/lib/dynamic_config/datastore_cache.rb

```diff
@@ -0,0 +1,93 @@
+# A caching layer that sits in front of a datastore that
+# implements get and set
+class DatastoreCache
+
+  class ConcurrentRefreshError < StandardError
+  end
+
+  # @param datastore [Object]
+  # @param cache_expiration [int] seconds after which a cached entry expires
+  def initialize(datastore, cache_expiration: 30)
+    @cache = {}
+    @datastore = datastore
+    @cache_expiration = cache_expiration
+
+    @global_lock = Mutex.new
```

Comment:
This is more an instance lock than a global lock (only applies to this database instance).  Perhaps call it @datastore_lock or just @luck?